File: WINGS_AI_Bootcamp/Haskell/kaprekarSeq.hs
import Data.List (sort)
import Control.Monad (when)

digits :: Int -> [Int]
digits 0 = []
digits x = digits (x `div` 10) ++ [x `mod` 10]

fromDigits :: [Int] -> Int
fromDigits = foldl (\acc d -> acc * 10 + d) 0

nextNumber :: Int -> Int
nextNumber n =
  let ds = digits n
      largest = fromDigits $ reverse (sort ds)
      smallest = fromDigits $ sort ds
  in largest - smallest

generate :: Int -> IO ()
generate n = do
  print n
  let next = nextNumber n
  when (next /= n) (generate next)

main :: IO ()
main = generate 1729

File: WINGS_AI_Bootcamp/Elixir/kaprekarSeq.ex
defmodule Sequence do
  def generate(n) do
    IO.puts(n)
    next = next_number(n)
    if next == n do
      :ok
    else
      generate(next)
    end
  end

  defp next_number(num) do
    digits = Integer.digits(num)
    largest = digits |> Enum.sort(&>=/2) |> Enum.join() |> String.to_integer()
    smallest = digits |> Enum.sort() |> Enum.join() |> String.to_integer()
    largest - smallest
  end
end

# Example usage:
Sequence.generate(1729)

